lib/open_api_spex/cast_parameters.ex

```diff
@@ -31,11 +31,16 @@ defmodule OpenApiSpex.CastParameters do
     Plug.Conn.fetch_cookies(conn).req_cookies
   end
 
+  # Matching is done using the downcase version of the expected_names but the original
+  # version of the expected_names is kept
   defp get_params_by_location(conn, :header, expected_names) do
+    expected_names = Map.new(expected_names, & {String.downcase(&1), &1})
+
     conn.req_headers
-    |> Enum.filter(fn {key, _value} ->
-      Enum.member?(expected_names, String.downcase(key))
+    |> Enum.map(fn {key, value} ->
+      {Map.get(expected_names, String.downcase(key)), value}
```

Comment:
`key` from `conn.req_headers` will always be lower-case, but I guess it doesn't hurt to be extra sure.